Added cache support, removed extra 'fs' calls for performance, etc.

dcombslinkedin opened the pull request:
Most code now in index.js.

Left code in lib that is still used externally (e.g. deps-for, resolve-pkg, pkg) that is
no longer needed for hash-for-dep itself.

stefanpenner commented:
@dcombslinkedin nice, but it looks like there are some new failing tests, mind fixing?

dcombslinkedin commented:
Thanks, Stef. As I haven't found a way to run checks for lint or es5 against the module (is there one?), it's been a one-at-a-time parade of fixes. Working on the latest update now.

stefanpenner commented:
> stefanpenner closed this 20 seconds ago

Hmm, i have no memory of this...

stefanpenner commented on index.js:
```diff
@@ -1,29 +1,255 @@
+ * may not be initialized until later (this then supports creating cycles
+ * in the dependencies list, which is supported by Node).
+ */
+var PackageCacheEntry = function(name, version, rootDir, sourceHash) {
```
nit-pick for consistency, can we use the following form ?

```js
function PackageCacheEntry(name, version, rootDir, sourceHash) {

}
```

dcombslinkedin replied:
Oh, certainly! Not sure why I did it the other way - I don't usually use that form.

stefanpenner commented on index.js:
```diff
@@ -1,29 +1,255 @@
+  var realPath;
+
+  try {
+    realPath = CACHES.REAL_FILE_PATH.get(filePath);
```
the cache check, likely should not be in the `try/catch`, I believe the `try/Catch` is focused on the IO/FS related errors.

dcombslinkedin replied:
Moved setting realPath and the check for non-null out of the try/catch

stefanpenner commented on index.js:
```diff
@@ -36,73 +262,238 @@ function cacheSet(key, value) {
+    // The keys of the returned object are the package baseDir values (the
+    // packageCacheEntry.rootDir). The values are the packageCacheEntry
+    // objects. 
+    var dependencies = Object.create({});
```
I believe `dependencies` here is a growing dictionary right? If so, I think `Object.create(null)` would be more appropriate, as that immediately creates a dictionary, whereas `Object.create({})` first creates a new possible FastMap that inherits from `{}`, but since the result continues to grow, eventually the VM will bail out and fallback to a dictionary anyways. Which seems wasteful.

dcombslinkedin replied:
Good point. I need to verify going to 'null', because there was a case (I don't think it was in this file, but I don't remember, it was a few weeks ago) where using 'null' meant that it didn't have access to a standard Object method that I was trying to use. The cache does grow, but only for enough time to get the dependencies for the original object transitively, then it's thrown out. Anyway, no objection to null, just let me verify that it actually doesn't trip over that other thing I found.

dcombslinkedin replied:
Verified that using null works, so marking this resolved.

stefanpenner commented:
@dcombslinkedin broadly this looks good, I left some nit-picks and touch ups. Do to the size of the change, I'll need to spend some extra time with it yet, I hope to have my review complete by EOD.

stefanpenner commented:
I've restarted the windows tests.

dcombslinkedin commented:
Thanks, Stef. I'm heading up to Redmond today for a MS conf on 1ES/1DS and some other build stuff tomorrow. I'll try to get changes in tonight after I arrive, so I should have something for you either tonight or Friday morning.

stefanpenner commented:
@dcombslinkedin I've fixed my above nit-picks, and forced push.

stefanpenner commented:
Add refactoring/extracting and writing tests

stefanpenner commented:
I've made the following changes:

* split the functions and algorithms into there own files for testing/reading
* caches are now passed as arguments to functions
* the outermost function `hashForDep` now selects between the global cache or a function allocated local cache.  This removes the need to manage the global state, instead allowing the GC to handle it all for us.

dcombslinkedin commented on index.js:
```diff
@@ -1,29 +1,255 @@
+ * @param  {String} filePath the path to the file of interest
+ * @return {String} real path or null
+ */
+function getRealFilePath(filePath) {
```
I'm fine with adding tests, but I'd really prefer that getRealFilePath, findPackagePath and resolvePackagePath live in a single file (separate from hashForDep is fine. They're all strongly related, and ultimately doing one job. It's harder to follow what's going on when they're spread across multiple files.

stefanpenner replied:
If you prefer to colocate them into a single utility file and export them that is god with me, my goals are:

* ensure these functions remain stateless
* ensure these functions are tested

Both of those can be accomplished in 1 file.

dcombslinkedin replied:
I moved getRealFilePath inside resolve-package-path as a private fn, _getRealFilePath, and have tests for it in the new file resolve-package-path-test.js. Similarly for findPackagePath. Removed the separate files for each of these two.

dcombslinkedin commented on index.js:
```diff
@@ -1,29 +1,255 @@
+ *
+ * @result the path to the actual package.json file that's found, or null if not.
+ */
+function findPackagePath(name, dir) {
```
see above comment

dcombslinkedin commented on index.js:
```diff
@@ -36,73 +262,238 @@ function cacheSet(key, value) {
+ * entries. The 'sourceHash' values are computed at the time the
+ * PackageCacheEntry is created, so don't need to be done again.
+ */
+function computeHash(packageCacheEntry, dependencies, heimdallNode) {
```
Sounds good. I thought the original hash checks were okay as you had them, but if you want to split them separately to this function, no issue.

stefanpenner commented on tests/module-entry-test.js:
```diff
@@ -0,0 +1,11 @@
+var assert = require('assert');
+var ModuleEntry = require('../lib/module-entry');
+
+describe.skip('ModuleEntry', function() {
```
@dcombslinkedin can you implement these tests

dcombslinkedin replied:
done (at least to the point I can close this).

stefanpenner commented on tests/get-real-file-path-test.js:
```diff
@@ -0,0 +1,11 @@
+var assert = require('assert');
+var getRealFilePath = require('../lib/get-real-file-path');
+
+describe.skip('getRealFilePath', function() {
```
@dcombslinkedin can you implement these tests

dcombslinkedin replied:
Done, inside resolve-package-path-test.js. Checking in shortly.

stefanpenner commented on tests/gen-name-dir-key-test.js:
```diff
@@ -0,0 +1,11 @@
+var assert = require('assert');
+var genNameDirKey = require('../lib/gen-name-dir-key');
+
+describe.skip('gen-name-dir-key-test', function() {
```
@dcombslinkedin can you implement these tests

dcombslinkedin replied:
Done, in resolve-package-path-test.js. Checking in shortly.

stefanpenner commented on tests/find-package-path-test.js:
```diff
@@ -0,0 +1,11 @@
+var assert = require('assert');
+var findPackagePath = require('../lib/find-package-path');
+
+describe.skip('findPackagePath', function() {
```
@dcombslinkedin can you implement these tests

dcombslinkedin replied:
Done (to the point I can resolve this conversation).

stefanpenner commented:
@dcombslinkedin after implementing the missing tests, can you investigate the windows failures?

dcombslinkedin commented on index.js:
```diff
@@ -36,73 +48,49 @@ function cacheSet(key, value) {
-  } else {
-    key = cacheKey(name, dir);
-  }
+  var caches = skipCache ? new CacheGroup() : PROCESS_CACHE;
```
Yep - good idea!

dcombslinkedin commented on index.js:
```diff
@@ -36,73 +48,49 @@ function cacheSet(key, value) {
-      CACHE.set(key, hash);
-    }
+  try {
+    return ModuleEntry.locate(caches, name, dir, hashTreeFn).getHash(h);
```
good deal.

dcombslinkedin commented on index.js:
```diff
@@ -1,30 +1,42 @@
+/*
+ * PROCESS_CACHE is the default cache used until the process exists
+ */
+var PROCESS_CACHE = new CacheGroup();
```
Makes good sense.

dcombslinkedin commented on lib/cache-group.js:
```diff
@@ -0,0 +1,35 @@
+var Cache = require('./cache');
+
+/*
+ * CacheGroup is used to both speed up, and ensure consistency of hashForDep.
```
Will do so tonight - thanks. I only saw a couple nits to take care of.

stefanpenner replied:
:)

dcombslinkedin replied:
Updated the docs as necessary, adding explanations where I thought they'd be helpful, verifying new names of things were correctly referenced, etc.

dcombslinkedin commented on lib/cache-group.js:
```diff
@@ -0,0 +1,35 @@
+ *    'fs.statSync' methods.
+ */
+
+module.exports = function CacheGroup() {
```
Yep. I thought about doing that, but hadn't taken care of it - thanks for the fix!

dcombslinkedin commented on lib/hash-tree.js:
```diff
@@ -33,7 +33,18 @@ function stringifyFileInfo(fileInfo) {
+
+  var buf;
+  // Once we drop old new support, we can simplify this code to Buffer.from
+  if (typeof Buffer.from === 'function') {
```
Better solution than mine - thanks!

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+ * in the dependencies list, which is supported by Node).
+ */
+module.exports = ModuleEntry;
+function ModuleEntry(name, version, rootDir, sourceHash) {
```
I was using 'package', but 'module' is better. No issue with the change at all.

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+  // the computed hash for the package's source files plus
+  // the source hashes of the dependencies. Because we compute
+  // dependencies as 'all the way down', this gets everything.
+  this._hash = null;
```
I wondered about that - thanks! I wasn't sure whether the value was supposed to be private or not, but have no issue if it is.

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+  this._dependencies = null;
+}
+
+ModuleEntry.prototype.initializeDependencies = function(caches, pkg, hashTreeFn) {
```
Nice change! thanks.

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+ * NOTE: for speed, the initial caller of gatherDependencies must pass
+ * in a result object, so we aren't constantly checking that it is set.
+ */
+ModuleEntry.prototype._gatherDependencies = function(dependencies) {
```
Very nice! I like this better than mine a lot - I hadn't done a "localize processing inside the object where possible" pass yet, though I should have. Thanks for the reminder!

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+      return caches.MODULE_ENTRY.get(realPathKey);
+    }
+
+    var thePackage = tryRequire(realPath);
```
Saw that. Nice!

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+    caches.MODULE_ENTRY.set(realPathKey, moduleEntry);
+    caches.PATH.set(nameDirKey, realPathKey);
+
+    moduleEntry.initializeDependencies(caches, thePackage, hashTreeFn);
```
good deal - no issues.

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+
+    // Insert the new ModuleEntry into the cache.MODULE_ENTRY so we know to stop
+    // when dealing with cycles.
+    var moduleEntry = new Constructor(thePackage.name, thePackage.version, rootDir, sourceHash);
```
Interesting - I'd never seen 'new Constructor' used before.

dcombslinkedin commented on lib/resolve-package-path.js:
```diff
@@ -0,0 +1,25 @@
+ * Canonicalize a file path from name, dir into a real path. It's possible the path
+ * does not actually resolve to a real file, in which case null is returned.
+ */
+module.exports = function resolvePackagePath(realFilePathCache, name, dir) {
```
I originally thought about doing that (non-global state), but at the time thought that it was not great to pass it around everywhere. Now that I see where you did it, I have no problem with it, and like that it removes the use of global state, specifically for the temp cache. Nicely done!

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+ * entries. The 'sourceHash' values are computed at the time the
+ * PackageCacheEntry is created, so don't need to be done again.
+ */
+ModuleEntry.prototype.getHash = function(heimdallNode) {
```
Good idea! thanks for the fix.

dcombslinkedin commented on index.js:
```diff
@@ -36,73 +48,49 @@ function cacheSet(key, value) {
-      CACHE.set(key, hash);
-    }
+  try {
+    return ModuleEntry.locate(caches, name, dir, hashTreeFn).getHash(h);
```
I've done a pass through the code to fix comments and do the co-location of those three functions (only one is now public, the resolvePackagePath one, but the others are private, named with _, and exported for testing. I've verified that that all seems to work, and done a live debug to see that the right stuff is called at the right places. I'll add the new tests and resolve windows issues (if there are any) tomorrow after the conf. at Microsoft. At the worst I should have something for you Friday morning. I hesitate to check things in right now just in case they're somehow bad, though I see no evidence of that right now.

rwjblue commented on lib/find-package-path.js:
```diff
@@ -0,0 +1,57 @@
+    var currPath = dir;
+
+    while (currPath !== fsRoot) {
+      var endsWithNodeModules = path.basename(currPath) === 'node_modules';
```
FWIW, I think we should be careful not to bake the `node_modules` assumption too deeply here. Not sure that we need to change anything in practice just yet, but under PnP `./node_modules` just doesn't exist...

dcombslinkedin replied:
PnP? Not sure what that is. The code here is basically the same as in ember-cli package-info-cache, but with the extra optimization that we are only looking for a package.json, not a directory. Can you expand a tad on what you mean?

rwjblue commented on lib/hash-tree.js:
```diff
@@ -33,7 +33,18 @@ function stringifyFileInfo(fileInfo) {
-  return crypto.createHash('md5').update(Buffer.from(strings, 'utf8')).digest('hex');
+
+  var buf;
+  // Once we drop old new support, we can simplify this code to Buffer.from
```
```suggestion
  // Once we drop Node < 6 support, we can simplify this code to Buffer.from
```

rwjblue commented on lib/hash-tree.js:
```diff
@@ -33,7 +33,18 @@ function stringifyFileInfo(fileInfo) {
+  if (typeof Buffer.from === 'function') {
+    buf = Buffer.from(strings);
+  } else {
+    // Old Node versions don't support Buffer.from, so use 'new' instead.
```
I don't think we need comments on both sides of this simple condition...

```suggestion
```

rwjblue commented on lib/hash-tree.js:
```diff
@@ -33,7 +33,18 @@ function stringifyFileInfo(fileInfo) {
+    buf = new Buffer(strings, 'utf8');
+
+  }
+  var hash = crypto.createHash('md5').update(buf).digest('hex');
```
Any reason not to return here directly?

dcombslinkedin replied:
Not really, though sometimes for debugging it's nice to put the value into a variable first so you can check it in the debugger. Do you guys have a preference in those sorts of situations?

dcombslinkedin replied:
Switched to just returning the expression value instead of assigning first.

rwjblue commented on lib/gen-name-dir-key.js:
```diff
@@ -0,0 +1,6 @@
+'use strict';
+var cacheKey = require('./cache-key');
+
+module.exports = function genNameDirKey(name, dir) {
```
I don't really grok why we need this file, but the implementation could be simplified to:

```suggestion
module.exports = cacheKey;
```

dcombslinkedin replied:
I originally had genNameDirKey and genModulePathKey (or whatever it was) in the index.js file. I abstracted away from 'cacheKey' because that name was so general and non-obvious. I'm a bit unclear on why we need to have cache keys instead of using the original strings for name+dir, path and realFilePath. Any background either of you can give me on why?

dcombslinkedin replied:
I ended up just going back to using 'cacheKey' in both places, eliminating the separate functions, and adding a short comment to cache-key.js. I still have the question about why we need to cache the key instead of the raw string, but am otherwise good to go.

stefanpenner commented on lib/cache.js:
```diff
@@ -2,13 +2,15 @@
 
 function makeCache() {
   // object with no prototype
+  // we still support some old node versions, eventually we can just drop this
```
Given that, we should likely just remove the hack all together, I don't think its worth keeping around.

rwjblue replied:
ya, I agree. IMHO its important to stay functional for older Node versions, but performance doesn't matter there....

dcombslinkedin replied:
Removed the hack - now either Buffer.from or new Buffer() if that doesn't exist.

rwjblue commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+  // in the package.dependencies field.) Later, hashForDep calls 'gatherDependencies'
+  // to compute the complete list including transitive dependencies, and that is then
+  // used for the final hash calculation.
+  Object.keys(pkg.dependencies).sort().forEach(function(dep) {
```
Should probably also care about `optionalDependencies`, right?

dcombslinkedin replied:
The original code didn't specifically mention optionalDependencies (I'd never heard of that property, just dependencies and devDependencies), but if they should be tracked for changes, agree with your point. Do you agree given the above comment?

stefanpenner commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+      return caches.MODULE_ENTRY.get(realPathKey);
+    }
+
+    var thePackage = tryRequire(realPath);
```
document why this is needed

dcombslinkedin replied:
Sorry, which 'this' do you mean - tryRequire?

stefanpenner replied:
yes `tryRequire`

dcombslinkedin replied:
Added the following comment: 
    // require the package.  If we get a 'module-not-found' error it will return null
    // and we can insert an entry in the cache saying we couldn't find the package
    // in case it's requested later. Any other more serious error we specifically
    // don't try to catch here.
If you think this is unnecessary or overkill, I'm okay with just doing a require w/o tryRequire, just let me know. We might want to put out a console message too - thoughts?

rwjblue commented on lib/resolve-package-path.js:
```diff
@@ -0,0 +1,25 @@
+  var h = heimdall.start('hashForDep:resolvePackagePath');
+
+  try {
+    if (name === './' || name.charAt(0) === '/') {
```
Do we normalize to posix paths? How does this work on windows?

dcombslinkedin replied:
I don't think so, don't know for sure. If it can be done fast and as few times as possible, great. If it has to be done for every path during every resolution, leaving them as is seems faster. But I'm willing to be convinced I'm wrong. :-)

stefanpenner commented on lib/cache-group.js:
```diff
@@ -0,0 +1,35 @@
+ *
+ * The CacheGroup contains 3 seperate caches:
+ *
+ *  - MODULE_ENTRY: the cache of PackageCacheEntry objects with information
```
can we surface the  cache key, and cache value of MODULE_ENTRY cache. Similarly to what you did for `PATH` bellow

dcombslinkedin replied:
Question for you: is there a reason that we're using a hash at all for these keys as opposed to just the actual string value? Just a shorter string? Something else? Let me take a look about what was done with PATH (I'm on the plane and having connection issues, so getting source files right now is an issue.)

dcombslinkedin replied:
Stef, I lost the context of what you mean above - where is the 'PATH below' now that you're talking about? Also, why do you want to surface the cache key and value of MODULE_ENTRY? What sort of use-case are you thinking of?

stefanpenner replied:
for `PATH` you have `nameDirKey => realPathKey.` which is really nice and descriptive, as to what the cached key/value pair are. 

Could you also surface that here for `MODULE_ENTRY`

dcombslinkedin replied:
Oh, so you just mean changing the var names to be more descriptive? Sure, happy to fix that.

dcombslinkedin replied:
Done. Will check in shortly.

stefanpenner commented on lib/cache-group.js:
```diff
@@ -0,0 +1,35 @@
+ *    realPathKey is a hash of the resulting real path to the relevant package.json file.
+ *    See the genRealPathKey() and genNameDirKey() functions.
+ *
+ *  - REAL_FILE_PATH: the cache of a file path to the resolved real path of that file.
```
Can we surface the cache key, and cache value of REAL_FILE_PATH cache. Similarly to what you did for PATH above.

dcombslinkedin replied:
I need to understand this, will look tomorrow morning.

stefanpenner replied:
for `PATH` you have `nameDirKey => realPathKey.` which is really nice and descriptive, as to what the cached key/value pair are. 

Could you also surface that here for `REAL_FILE_PATH `

stefanpenner commented on lib/cache-key.js:
```diff
@@ -1,8 +1,14 @@
-  var value = name + 0x00 + dir;
-
-  return crypto.createHash('sha1').update(value).digest('hex');
+  return crypto.createHash('sha1').update(name + '|' + dir).digest('hex');
```
should we use `"\x00"` instead of '|' ? So we have a null char, and no accidental collisions?

dcombslinkedin replied:
I don't think so, because 0x00 is not a null character, it's converted from numeric 0 to a '0' string when inserted into the string. That's much worse than |, IMO.

stefanpenner replied:
Ya, i typo'd it by mistake [more](https://github.com/stefanpenner/hash-for-dep/pull/45#discussion_r)

dcombslinkedin replied:
I switched to using '\x00'.

stefanpenner commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+
+  var sourceHashes = dependencyRootDirs.map(function(rootDir) {
+    return dependencies[rootDir]._sourceHash;
+  }).join('|');
```
should we use `"\x00"` instead of '|' ? So we have a null char, and no accidental collisions?

dcombslinkedin replied:
See earlier reply - 0x00 doesn't do what you think it does when + into a string.

stefanpenner replied:
I believe it does `"\x00".charCodeAt(0) === 0` 

I believe i simply made a typo earlier and did `0x00` instead of `\x00`

dcombslinkedin replied:
Ah, got it. Let me test, and if it's working, I'm fine with using it. It did always seem kind of odd. :-)

dcombslinkedin replied:
I switched to using \x00 except in hash-tree (I think you were using that already, right?) If you want it there, too, fine with me.

dcombslinkedin commented on index.js:
```diff
@@ -36,73 +48,49 @@ function cacheSet(key, value) {
 module.exports = function hashForDep(name, dir, _hashTreeOverride, _skipCache) {
-  var skipCache = false;
-  var key, hash;
+  var skipCache = (typeof _hashTreeOverride === 'function' || _skipCache === true) ? true : false;
```
I was never sure myself - it was in the code when I got hold of it.

dcombslinkedin replied:
Actually that's wrong - the first part to assign 'true' to it was, the second part skipped on to create something else. I've removed the ternary, as you're right, it's not necessary. Coming in the next checkin.

dcombslinkedin commented on lib/find-package-path.js:
```diff
@@ -0,0 +1,57 @@
+  var h = heimdall.start('hashForDep:findPackagePath');
+
+  try {
+    // Early versions of Node don't support 'path.parse', and Windows
```
Will change when I get in tomorrow morning. The only reason I used path.parse was to get the root dir, so not sure it's really necessary to pull it in vs the manual implementation. I assume that path.parse, when implemented, is still more expensive either way. But I'm happy to be convinced if you think it will really buy us something.

dcombslinkedin commented on lib/find-package-path.js:
```diff
@@ -0,0 +1,57 @@
+  try {
+    // Early versions of Node don't support 'path.parse', and Windows
+    // doesn't use '/', so here we go.
+    var fsRoot = (IS_WINDOWS ? dir.split(path.sep)[0] : '/');
```
Since it's only for getting the root, path-root is better. I thought a straightforward implementation was "better", but agree it's something of a hack. Let me check on path-root in the morning.

dcombslinkedin commented on index.js:
```diff
@@ -1,29 +1,255 @@
+ * Canonicalize a file path from name, dir into a real path. It's possible the path
+ * does not actually resolve to a real file, in which case NO_PATH_FOUND is returned.
+ */
+function resolvePackagePath(name, dir) {
```
Done locally, will go into the next checkin on my branch.

dcombslinkedin commented on index.js:
```diff
@@ -36,73 +262,238 @@ function cacheSet(key, value) {
+ * node_modules dependencies) at the same time to make it easier to establish
+ * the 'hash' value (includes dependency sourceHash values) later.
+ */
+function computePackageInfo(name, dir, hashTreeFn) {
```
Stef moved to ModuleEntry.locate() in module-entry.js

dcombslinkedin commented on lib/module-entry.js:
```diff
@@ -0,0 +1,227 @@
+ * the same time to make it easier to establish the 'hash' value (includes
+ * dependency sourceHash values) later.
+ */
+ModuleEntry.locate = function(caches, name, dir, hashTreeFn) {
```
'locate' sounds a tad odd, but I have no other better suggestion, so am happy to leave it as is. And the positioning as an instance method totally makes sense.

dcombslinkedin commented:
I'm making a few changes in addition to addressing the items above. New changes include:

- Cleaned up unused imports from index.js
- Converted 'genNameDirKey' and 'genRealPathKey' with direct references to cacheKey, and added a comment to cacheKey. It seems overkill to have separate files and test files for what are basically one-line wrappers around the cacheKey calls. I'm happy to add tests to the existing cache-key tests if that's necessary/useful. As part of this, removed the gen-name-dir-key.js and tests/gen-name-dir-key-test.js.

stefanpenner commented on lib/resolve-package-path.js:
```diff
@@ -1,12 +1,115 @@
+      var endsWithNodeModules = path.basename(currPath) === 'node_modules';
+
+      var nodeModulesPath = endsWithNodeModules ? currPath : currPath + path.sep + 'node_modules';
+      var filePath = path.join(nodeModulesPath, name);
```
Question, since we are already bypassing `path.join` in the line above, should we here as all?

`var filePath  = nodeModulesPath + path.sep + name`

Or should we use `path.join` in both cases.

dcombslinkedin replied:
Though it's slower (being a function), I think I favor path.join() in both cases. Thoughts?

dcombslinkedin replied:
Ended up going with path.join in both.

stefanpenner commented on lib/module-entry.js:
```diff
@@ -0,0 +1,237 @@
+    // We have the package object and the relevant keys.
+    var rootDir = realPath.slice(0, realPath.length - 12 /* index of `/package.json` */);
+
+    console.log("realPath rootDir = " + rootDir);
```
I suspect this was leftover from earlier debugging, I believe it should be removed.

dcombslinkedin replied:
Yes. Just in to help with Windows debugging and will be removed.

stefanpenner commented on lib/resolve-package-path.js:
```diff
@@ -0,0 +1,137 @@
+    while (currPath !== fsRoot) {
+      var endsWithNodeModules = path.basename(currPath) === 'node_modules';
+
+      var filePath = (endsWithNodeModules ? path.join(currPath, name) : path.join(currPath, 'node_modules', name));
```
In what scenarios would the module `name` not be found at `<some-path>/node_modules/name`?

If so, could the alg not be improved by skipping all paths who's basename is not node_modules ?

stefanpenner replied:
Interestingly if `endsWithNodeModules` is always `false` all the tests pass.

stefanpenner replied:
Ya the algorithm as written right seems to do duplicate IO in some scenarios.

```
'/Users/spenner/src/stefanpenner/hash-for-dep/tests/fixtures/node_modules/foo/node_modules/bar/node_modules/foo/package.json'
'/Users/spenner/src/stefanpenner/hash-for-dep/tests/fixtures/node_modules/foo/node_modules/foo/package.json'
'/Users/spenner/src/stefanpenner/hash-for-dep/tests/fixtures/node_modules/foo/node_modules/foo/package.json'
'/Users/spenner/src/stefanpenner/hash-for-dep/tests/fixtures/node_modules/foo/package.json'
```

stefanpenner replied:
if `endsWithNodeModules` was true, we need to decrement the `currPath` twice, other wise the next iteration of the loop we will check the same directory again.

dcombslinkedin replied:
> Interestingly if `endsWithNodeModules` is always `false` all the tests pass.

I definitely want to look into this. I've removed the change I made in this version (including the double dirname popping) and will go through it again to make sure it all works.

dcombslinkedin replied:
reworked.

stefanpenner commented on lib/resolve-package-path.js:
```diff
@@ -0,0 +1,137 @@
+    var currPath = dir;
+
+    while (currPath !== fsRoot) {
+      var endsWithNodeModules = path.basename(currPath) === 'node_modules';
```
interestingly `node_modules` here must actually be case-insensitive, as the directory on disk may be fully or partially camalized.

Given:
```
/NODE_MODULES/a/package.json
```

running
```
node -e "require('a/package.json');
```

Works

dcombslinkedin replied:
good catch - thanks.  Will address today.

dcombslinkedin replied:
Done. Checking in shortly.

dcombslinkedin replied:
Added a 'toLowerCase() to the path.basename() value.

stefanpenner commented on lib/resolve-package-path.js:
```diff
@@ -0,0 +1,170 @@
+
+      // pop to next higher node_modules. This assumes that
+      // node_modules/node_modules is not possible.
+      currPath = path.dirname(currPath);
```
I don't believe this is correct as per my earlier recommendation, the double "pop" of the current directory should only occur if the `currPath` ends in node_modules.

see: https://github.com/stefanpenner/hash-for-dep/pull/45/files#r254023431


Also, node_modules can occur at ever directory level not just the first one. Given:

`findPackagePath(cache, 'apple', '/foo/bar/baz/node_modules/')`

I would expect the following locations be considered as a location for `apple`

```
'/foo/node_modules/bar/orange/baz/node_modules/apple'
'/foo/bar/orange/node_modules/apple'
'/foo/bar/node_modules/apple'
'/foo/node_modules/apple'
'/node_modules/apple'
```

dcombslinkedin replied:
yep, that's where it should go. The change I put in is wrong here. Will fix today.

dcombslinkedin commented on lib/resolve-package-path.js:
```diff
@@ -1,12 +1,115 @@
+  try {
+    var stat = fs.statSync(filePath);
+
+    if (stat.isFile() || stat.isFIFO()) {
```
added comment to _getRealFilePath.

dcombslinkedin commented on lib/resolve-package-path.js:
```diff
@@ -1,12 +1,115 @@
+function _getRealFilePath(cache, filePath) {
+  var realPath = cache.get(filePath);
+
+  if (realPath) {
```
Changed the code to include the cache.has test first. Good catch, Stef!

dcombslinkedin commented on index.js:
```diff
@@ -36,73 +45,50 @@ function cacheSet(key, value) {
 module.exports = function hashForDep(name, dir, _hashTreeOverride, _skipCache) {
-  var skipCache = false;
-  var key, hash;
+  console.log("### calling hashForDep for name '" + name + "', dir '" + dir + "'");
```
Yes, we don't want to leave in any console.log() statements. I was just using it as a way of helping debug the windows issues quickly. When I get through that, I'll remove all of these.

stefanpenner commented:
FYI: https://github.com/stefanpenner/hash-for-dep/pull/48

stefanpenner commented:
looks like windows is green!

dcombslinkedin commented:
Yeah, the change to force core.symlinks to true seems to have done it. I'll remove the debug stuff and  probably add a couple more tests and push one more time for your (hopefully last) review.

dcombslinkedin commented:
Ignore this if it passes - forgot an 'it.only' and to squash things. Doing that now.

dcombslinkedin commented:
I'm going to try to squash, but have been having issues with it. Wish me luck!

stefanpenner commented:
released as v1.4.0 🎉
